Node revamp: PostgreSQL lessons (#27960)
TheOdinProject/curriculum: nodeJS/express/installing_postgresql.md

> <!-- markdownlint-disable MD024 TOP004 -->

Reviewer: I feel like we shouldn't disable rules as a temporary workaround to avoiding check failures, but rather for when we actually intend to disable a rule for good for whatever reason. I'd personally rather merge something that we know fails the linter, but intend to resolve as a followup. @MaoShizhong wdyt?

Reviewer 2: Perhaps follow the Rails version with a todo comment? Not the most ideal but it's at least consistent 

```suggestion
<!-- TODO: Revisit lesson/heading structure to remove need to disable rules -->
<!-- markdownlint-disable MD024 TOP004 -->
```

Reviewer: That sounds good

Author: above change made.

### Introduction

Data persistence is integral to the vast majority of web applications. Data persistence is achieved through databases. Being able to understand how to structure, build, and query your own database are important skills for any full-stack developer to have.

For example, consider The Odin Project. Where is your username stored? Where are project submissions stored? Where does this lesson content, the one you're reading right now, get stored? Yes, a database!

We've chosen [PostgreSQL](https://www.postgresql.org/) as our database of choice for the curriculum. All SQL knowledge that you've learned in the earlier course should be transferable to PostgreSQL. We'll also learn new PostgreSQL specific things. Watch through this [fun and short introduction video to PostgreSQL by Fireship](https://www.youtube.com/watch?v=n2Fluyr3lbc).

<div class="lesson-note lesson-note--critical" markdown="1">

> #### Prerequisite SQL knowledge

Reviewer: Is this note box necessary? If this lesson is placed after the SQL course it seems unnecessary to tell users that they should have done some other course first.

If it ultimately is necessary, I'd opt to removing this header and instead just have the content of the note box speak for itself. I don't think we would really *need* a separate heading for this.

Author: I think it's necessary. When we release the new updates, I'd expect _some_ people to jump straight into the new lessons in the Node course without going through the newly added SQL course. 

Sure, I'll remove the headings.

Make sure you've completed the [SQL course](https://www.theodinproject.com/paths/full-stack-javascript/courses/databases).

This lesson and all subsequent lessons will assume you understand SQL syntax and concepts.

</div>

> In this lesson, we'll install PostgreSQl and in the next lesson we'll learn how to use it in our Express application.

Reviewer 2: ```suggestion
In this lesson, we'll install PostgreSQL and in the next lesson we'll learn how to use it in our Express application.
```

### Lesson overview

This section contains a general overview of topics that you will learn in this lesson.

- How to install PostgreSQL.

### Installing PostgreSQL

Pick your operating system below to get the appropriate steps for installing PostgreSQL.

<details markdown="block">

> <summary class="dropDown-header">Linux</summary>

Reviewer: @MaoShizhong From your comment on the other PR, yeah the way we're doing these details/summary elements isn't great. Right now the heading structure from an AT perspective is essentially:

- Installing PostgresSQL
  - Step 1: Make sure the system is up to date
  - ....other headings....
  - Step 1: Make sure the system is up to date

Without any context of an OS anywhere. We could put this in as is as I'm not 100% sure how best to go about it yet. We could have the actual steps be visible at all times, then in each step have a details/summary per OS, so something like the following screenshot:

![image](https://github.com/TheOdinProject/curriculum/assets/70952936/e3086995-1427-4c13-84ec-23552b92f02c)

Since the content of the `summary` element can't be given a Markdown heading (we could use HTML `h3` etc, but then we need to allow that in the lint rules). But I'd have to stew on it a bit more probably.

Author: @thatblindgeye I straight up copy pasted the existing lesson in the Rails path in to a new file in this PR. If we do intend to make these changes, might I suggest handling it in a separate PR?

Reviewer: @01zulfi Oh no this is definitely something we're doing in other places, I hope my comment didn't come off as saying this was an issue you introduced or anything.

But yeah, this is something that can be a followup since it is something that will affect other lessons and there's still some ruminating to do on the idea.

Author: All good 😆 

I'm suggesting to open a separate issue/PR because this issue is relevant to other installation-esque lessons as well. Anyway, I'm happy to update this PR if we do reach a consensus on how to fix this issue.

Reviewer 2: I vote separate PR to handle both this and the Rails version together once decided on the best way to handle this.
Single headings each with OS-specific dropdowns does look appealing, though IIRC @thatblindgeye, I think you mentioned accessibility issues with the current actual method of doing the dropdown itself? Since if we can find an accessible and logical alternative that also conforms to the heading issue, we'd be wanting to update any other lesson that does things like this, like any of the earlier installations lessons

Reviewer: Yeah this all can be a followup

> ### Step 1: Make sure the system is up to date

Reviewer: The headings in these steps would need to be increased by one level since these are subsections of the larger "Installing PostgreSQL" section. Right now it would come off as this Step 1 section being more separate from "Installing" when it really isn't.

Author: (see above comment)

<!-- markdownlint-disable MD024 TOP004 -->

### Introduction

Data persistence is integral to the vast majority of web applications. Data persistence is achieved through databases. Being able to understand how to structure, build, and query your own database are important skills for any full-stack developer to have.

> For example, consider The Odin Project. Where is your username stored? Where are project submissions stored? Where does this lesson content, the one you're reading right now, stored? Yes, a database!

Reviewer 2: ```suggestion
For example, consider The Odin Project. Where is your username stored? Where are project submissions stored? Where does this lesson content, the one you're reading right now, get stored? Yes, a database!
```

> We've chosen [PostgreSQL](https://www.postgresql.org/) as our database of choice for the curriculum. All SQL knowledge that you've learned in the earlier course should be transferable to PostgreSQL. We'll also learn new PostgreSQl specific things. Watch through this [fun and short introduction video to PostgreSQL by Fireship](https://www.youtube.com/watch?v=n2Fluyr3lbc).

Reviewer 2: Here and a few other places in this lesson have `PostgreSQl` instead of `PostgreSQL` (lowercase l)

```suggestion
We've chosen [PostgreSQL](https://www.postgresql.org/) as our database of choice for the curriculum. All SQL knowledge that you've learned in the earlier course should be transferable to PostgreSQL. We'll also learn new PostgreSQL specific things. Watch through this [fun and short introduction video to PostgreSQL by Fireship](https://www.youtube.com/watch?v=n2Fluyr3lbc).
```

<div class="lesson-note lesson-note--critical" markdown="1">

#### Prerequisite SQL knowledge

Make sure you've completed the [SQL course](https://www.theodinproject.com/paths/full-stack-javascript/courses/databases).

This lesson and all subsequent lessons will assume you understand SQL syntax and concepts.

</div>

In this lesson, we'll install PostgreSQl and in the next lesson we'll learn how to use it in our Express application.

### Lesson overview

This section contains a general overview of topics that you will learn in this lesson.

- How to install PostgreSQL.

### Installing PostgreSQL

> If you've been following the curriculum, you should already have installed all the supporting bits and pieces. If not, please go back and refer to [those lessons](https://www.theodinproject.com/guides/installations) before continuing.

Reviewer 2: Do we need this paragraph? It's in the Rails version of this lesson but I've got a [note in my descriptive links spreadsheet for the Rails pathway](https://docs.google.com/spreadsheets/d/1rM9tjPIFZbKoMfd6OlLnjOhddBe2bivllY8_Bg5aGqY/edit#gid=1290166870&range=F30) that questions its value, given this curriculum is supposed to be taken linearly. I guess this probably also applies to things like the prerequisite boxes in both lessons.

At the very least, if we decide this paragraph is actually necessary, we'd need to change the link text to be more descriptive than "those lessons" (which will of course be changed accordingly in the Rails' equivalent when we get to that epic).

Author: agreed. removed the paragraph.

Pick your operating system below to get the appropriate steps for installing PostgreSQL.

<details markdown="block">

<summary class="dropDown-header">Linux</summary>

### Step 1: Make sure the system is up to date

Before installing PostgreSQL, it's a good idea to make sure the operating system is up to date. To update our system, run this command:

```bash
sudo apt update && sudo apt upgrade
```

### Step 2: Install the PostgreSQL packages

After our system is up to date, we will install the packages for PostgreSQL.

```bash
sudo apt install postgresql postgresql-contrib libpq-dev
```

After installation is complete, let's start the server using this command:

```bash
sudo systemctl start postgresql.service
```

<div class="lesson-note lesson-note--warning" markdown="1">

#### Systemctl and WSL2

Systemctl is not supported on WSL2, and the above command won't work. Instead, run the following command:

```bash
sudo service postgresql start
```

</div>

> Got an error, or don't see an active service? Come visit the [Discord](https://discord.gg/V75WSQG) for some help!

Reviewer 2: Same thing for line 196 in the MacOS instructions.
Planned descriptive link change to match the Rails version when that gets sorted.

```suggestion
Got an error, or don't see an active service? Come visit [our Discord server](https://discord.gg/V75WSQG) for some help!
```

#### 3.3 Creating the role database

One other important step in setting up PostgreSQL is that each role must have its own database of the same name. Without it, the role we just created will not be able to log in or interact with PostgreSQL.

You can try to run `psql` now, but you will get an error that the database does not exist. Not to worry, let's create one to resolve fix this:

<div class="lesson-note" markdown="1">

If your username has any capital letters, you must surround it in quotes when running the below command.

</div>

```bash
sudo -i -u postgres createdb <linux_username>
```

Now our role is fully set up: we've got `<role_name>` and that role has a database.

#### 3.4 Securing our new role

One important thing that we have to do is to set up a password for our new role so that the data is protected. Now that our role is set up, we can actually use it to administer PostgreSQL. All you have to do is enter this command to get into the PostgreSQL prompt:

```bash
psql
```

You should see the PostgreSQL prompt come up with the new role we just created, like so:

```sql
role_name=#
```

> If you don't see a similar prompt, then reach out on [Discord](https://discord.gg/V75WSQG) for some help. If you **do** see a similar prompt, then we can create a password for the role like so:

Reviewer 2: Same thing here (and for line 253)

```suggestion
If you don't see a similar prompt, then reach out in [our Discord server](https://discord.gg/V75WSQG) for some help. If you **do** see a similar prompt, then we can create a password for the role like so:
```
